NETOBSERV-2564: Add AI agents best practices documentation (#1239)
netobserv/network-observability-console-plugin: AGENTS.md

## Project Context

**NetObserv Web Console** - a network observability console that can be deployed either as:
- OpenShift console plugin: Integrates into OpenShift Console with extended views
- Standalone console: Single-page application showing Netflow Traffic and Network Health

**Components:**
- **Frontend**: TypeScript/React with PatternFly components (uses OpenShift Console dynamic plugin SDK for plugin mode)
- **Backend**: Go HTTP server providing API endpoints for Loki queries, Kubernetes resources, and Prometheus metrics

**Deployment Modes:**
- **Plugin mode**: Integrated into OpenShift Console
- **Standalone mode**: Independent web application

**Key Directories:**
- `web/src/components/`: React components (forms, tables, topology, etc.)
- `web/src/api/`: Backend API client
- `web/src/model/`: Data models, transformations, and context providers
- `web/src/utils/`: Utility functions and custom React hooks
- `web/src/standalone/`: Standalone mode application entry point
- `pkg/handler/`: HTTP request handlers
- `pkg/loki/`: Loki client and query builders
- `pkg/kubernetes/`: Kubernetes API client
- `pkg/prometheus/`: Prometheus client

> - `web/cypress/e2e/`: Cypress E2E tests (developer tests)

Reviewer: ```suggestion
- `web/cypress/e2e/`: Cypress E2E tests on mock data
- ```

> - `web/cypress/integration-tests/`: QE integration tests

Reviewer: ```suggestion
- `web/cypress/integration-tests/`: integration tests with openshift
```

- FLAVOR=`enduser` limits production standalone to Network Traffic/Health tabs

### Loki Query Optimization
- Limit time ranges, prefer `=` over `=~` matchers, push aggregations to LogQL
- Check `/api/loki/config/limits` for query constraints
- Mock mode: `make start-standalone-mock` or `make serve-mock`

### Frontend Configuration
Two types:
- **Dynamic (Operator-Generated)**: Runtime configuration from FlowCollector CR
  - Operator generates ConfigMap with columns, filters, scopes, features, port naming, quick filters
  - Generated from [static-frontend-config.yaml](https://github.com/netobserv/netobserv-operator/blob/main/internal/controller/consoleplugin/config/static-frontend-config.yaml) (operator repo) + FlowCollector spec
  - **Important**: Changes to `config/sample-config.yaml` frontend section must be synced to operator's `static-frontend-config.yaml`
  - Fetched via `/api/frontend-config`
- **Static (Development)**: Local configuration for development
  - `config/sample-config.yaml` for local development
  - Not used in production (operator generates config instead)

### PatternFly Components

> - Version: v5.x (see `web/package.json`)

Reviewer: ```suggestion
- Version: v6.x (see `web/package.json`)
```

Author: Need the PF6 PR to merge to make these changes. Will update once its merged

> - Follow PatternFly patterns, test light/dark themes (`pf-v5-theme-dark`)

Reviewer: need to update for pf6.

- Docs: [patternfly.org](https://www.patternfly.org/)

### Multi-Architecture
Support: amd64, arm64, ppc64le, s390x. Frontend built once, backend per arch. Build: `MULTIARCH_TARGETS="amd64 arm64" make images`

## Code Review Checklist

```text
Review for:
1. TypeScript type safety (no 'any' types without justification)
2. PatternFly component usage consistency
3. i18n strings for all user-facing text
4. Error handling (proper error messages and patterns)
5. Unit tests (React Testing Library for frontend, Go tests for backend)
6. Loki query efficiency (time ranges, label matchers)
7. Backward compatibility (schema changes)
8. Both plugin and standalone mode testing
9. Accessibility (ARIA labels, keyboard navigation)
10. Performance (avoid unnecessary re-renders, optimize queries)
11. Usability and user experience (intuitive workflows, clear feedback)
```

## Testing

**Unit Tests**: Go tests in `pkg/*_test.go`, Jest with React Testing Library in `web/src/**/__tests__/`

> **E2E Tests**: Cypress in `web/cypress/e2e/` (dev) and `web/cypress/integration-tests/` (QE). Run: `make cypress` or `cd web && npm run cypress:open`

Reviewer: ```suggestion
**E2E Tests**: Cypress in `web/cypress/e2e/` based on mock data, Run: `make cypress` or `cd web && npm run cypress:open`
**Integration tests with OpenShift**: `web/cypress/integration-tests/` (QE) will need a live Openshift cluster with version 4.19+
```

# AI Agents Best Practices for NetObserv Web Console

Best practices for AI coding agents on NetObserv Web Console.

> **Note**: Symlinked as [CLAUDE.md](CLAUDE.md) for Claude Code auto-loading.

## Project Context

**NetObserv Web Console** - OpenShift Console dynamic plugin for network observability visualization (also deployable as standalone app).

**Stack:**
- **Frontend**: TypeScript, React 18, PatternFly 6, React Router 7, i18next 25
- **Plugin SDK**: `@openshift-console/dynamic-plugin-sdk` 4.22+ (plugin mode only)
- **Backend**: Go HTTP server for Loki queries, Kubernetes resources, Prometheus metrics

**Deployment Modes:**
- **Plugin**: Integrated into OpenShift Console (OCP 4.22+)
- **Standalone**: Independent web application

**PatternFly Version Support Matrix:**
| OCP Version | Branch | PatternFly |
|-------------|--------|------------|
| ≥ 4.22 | `main` | PF6 |
| 4.19-4.21 | `main` or `main-pf5` | PF6 or PF5 |
| 4.15-4.18 | `main-pf5` | PF5 |

> | ≤ 4.14 | `main-pf4` | PF4 |

Reviewer 2: We should point the official doc https://github.com/openshift/console/tree/main/frontend/packages/console-dynamic-plugin-sdk#patternfly

Author: added link to patternfly doc

- `pkg/prometheus/`: Prometheus client
- `web/cypress/e2e/`: Cypress E2E tests on mock data
- `web/cypress/integration-tests/`: Integration tests with OpenShift

## Critical Constraints

### 🚨 OpenShift Console Plugin SDK
- Plugin mode must use `@openshift-console/dynamic-plugin-sdk` APIs
- Plugin mode must follow OpenShift Console conventions for navigation, extensions, theming
- Standalone mode uses the same codebase but without Console integration
- Test both plugin and standalone modes

### 🚨 Node Version
- Node.js 24 (source of truth: `Dockerfile.front`)
- `.npmrc` contains `legacy-peer-deps=true` for dependency resolution
- Use [nvm](https://github.com/nvm-sh/nvm) for version management

### 🚨 Dockerfiles
- `Dockerfile`: Production multi-stage build (frontend + backend)
- `Dockerfile.front`: Frontend build (Node.js version source of truth)
- `Dockerfile.downstream`: Downstream/productized builds
- `Dockerfile.ci`: CI environment builds
- `Dockerfile.e2e`, `Dockerfile.cypress`: Test runner images

### 🚨 Backward Compatibility
Frontend configuration schemas must remain compatible:
- ✅ Add optional fields with defaults
- ❌ Never remove/rename fields in schemas or API responses

> - Changes to `web/src/components/forms/config/uiSchema.ts` must preserve existing field behaviors

Reviewer 2: I'm wondering if this is true for console plugin. It is true for the operator but on the frontend side, anything can be changed actually.

Author: removed this section

- **Backend**: Go HTTP server providing API endpoints for Loki queries, Kubernetes resources, and Prometheus metrics

**Deployment Modes:**
- **Plugin mode**: Integrated into OpenShift Console
- **Standalone mode**: Independent web application

**Key Directories:**
- `web/src/components/`: React components (forms, tables, topology, etc.)
- `web/src/api/`: Backend API client
- `web/src/model/`: Data models and transformations
- `web/src/utils/`: Utility functions
- `pkg/handler/`: HTTP request handlers
- `pkg/loki/`: Loki client and query builders
- `pkg/kubernetes/`: Kubernetes API client
- `pkg/prometheus/`: Prometheus client
- `web/cypress/e2e/`: Cypress integration tests

## Critical Constraints

### 🚨 OpenShift Console Plugin SDK
- Must use `@openshift-console/dynamic-plugin-sdk` APIs
- Follow OpenShift Console conventions for navigation, extensions, theming
- Test both plugin and standalone modes

### 🚨 Node Version Consistency
- Use Node.js version specified in `Dockerfile.front`
- npm versions matter - use expected versions to avoid build breaks
- Consider using [nvm](https://github.com/nvm-sh/nvm) for version management

> 

Reviewer: We could add section for various Dockerfiles
